File: JULY-2025/28-07-2025/p2044.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countMaxOrSubsets = function(nums) {
    const n = nums.length;
    let max = nums.reduce((a, e) => a | e, 0);
    let count = 0;

    const dfs = (i, or) => {
        if (or === max) {
            // Optimization: if OR already equals max, we skip the rest of recursion.
            // The number of remaining subsets from here is 2^(nums.length - i)
            count += 1 << ( n - i);
            return;
        }

        if (i === n) {
            return;
        }

        dfs(i + 1, or);            // we exclude nums[i]
        dfs(i + 1, or | nums[i]);  // we include nums[i]
    };

    dfs(0, 0);
    return count;
};

File: JULY-2025/28-07-2025/p2044.rb
# @param {Integer[]} nums
# @return {Integer}
def count_max_or_subsets(nums)
    dp = Hash.new(0)
    dp[0] = 1
    nums.each do |nums|
      cl = dp.clone
      cl.each do |k, v|
        dp[k | nums] += v
      end
    end
    dp[dp.keys.max]
end